### PR TITLE
check finder type before passing path

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Attempt to fix https://github.com/ansible/ansible/pull/76225#issuecomment-984448424

##### ISSUE TYPE
- Bugfix Pull Request

### CODE REVIEW

**Review by nitzmahone — APPROVED** [review]
Fix looks sane, I've got some ideas for how to test...

**Comment by nitzmahone** [comment]
Something didn't feel right here- the fix is fine, but looking at current `pkg_resources`, I think this fix is actually masking a different problem: https://github.com/pypa/setuptools/blob/main/pkg_resources/__init__.py#L2207-L2213

At first glance, this is actually a setuptools bug that was introduced here: https://github.com/pypa/setuptools/pull/1563 - they added support for `find_spec`, which is allowed to return `None`, but `pkg_resources._handle_ns` assumes it'll return a spec and blindly dots into the result to get a loader (where the old path got the loader directly, which could be `None`). The `AttributeError` that's supposed to trigger py2 compatibility (because py2 loaders don't implement `find_spec`) is inadvertently triggered by dotting into `None`, so it's falling back to the py2 loader method, which (now that it's patched to work under py3) returns `None`, allowing the code to work. I think this is probably OK short-term, but might be worth a patch to setuptools, because if we didn't have a loader with py2 fallback methods (as was the original plan), this would just fail.

The setuptools patch is pretty trivial...

**Comment by nitzmahone** [comment]
Proposed a setuptools patch for this going forward, but we should also keep `find_module` working under 2.x in the meantime: https://github.com/pypa/setuptools/pull/2918

**Comment by nitzmahone** [comment]
FYI @felixfontein re: c.g test failures from the collection loader py3 update

File: lib/ansible/utils/collection_loader/_collection_finder.py [hunk 1]
```diff
@@ -299,23 +307,26 @@ def _get_finder(self, fullname):
     def find_module(self, fullname, path=None):
         # we ignore the passed in path here- use what we got from the path hook init
         finder = self._get_finder(fullname)
-        if finder is not None:
-            return finder.find_module(fullname, path=[self._pathctx])
-        else:
+
+        if finder is None:
             return None
+        elif HAS_FILE_FINDER and isinstance(finder, FileFinder):
```
Comment (s-hertel) [inline lib/ansible/utils/collection_loader/_collection_finder.py:313]:
```suggestion
        elif HAS_FILE_FINDER and isinstance(finder, FileFinder):
            # this codepath is erroneously used under some cases in py3,
            # and the find_module method on FileFinder does not accept the path arg
            # see https://github.com/pypa/setuptools/pull/2918
```

**Comment by felixfontein** [comment]
@s-hertel @nitzmahone thanks for fixing this!

**Pull request merged**